### PR TITLE
Add sized regions

### PR DESCRIPTION
**Pull request by davidtgoldblatt**
Third time's the charm. This implements the ideas in #887, and supersedes #951.

I feel much more confident in the overall strategy here, as I now am much more familiar with the extent module internals.

### CODE REVIEW

File: include/jemalloc/internal/sized_region.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+	JEMALLOC_ALIGNED(CACHELINE)
+	atomic_zu_t size_class_space_used[SIZED_REGION_NUM_SCS];
+};
+
+/* Whether this feature is enabled. */
+extern bool opt_sized_regions;
+/* The lg of the number of bytes reserved for each size class in the region. */
+extern ssize_t opt_lg_sized_region_size;
+/*
+ * The region we actually allocate out of.  There should be only of these per
```
Comment (djwatson) [inline include/jemalloc/internal/sized_region.h]: you missed a word

File: include/jemalloc/internal/arena_structs_b.h [hunk 2]
```diff
@@ -122,13 +122,22 @@ struct arena_bin_s {
...
 	/*
-	 * Heap of non-full slabs.  This heap is used to assure that new
+	 * Heaps of non-full slabs.  These heaps are used to assure that new
 	 * allocations come from the non-full slab that is oldest/lowest in
-	 * memory.
+	 * memory.  The _sized heap contains extents from the sized alloc
+	 * region, whereas the _unsized one contains all others.  We prefer the
+	 * sized-region ones over the unsized ones where we can.
+	 *
+	 * Note that we could get by by simply preferring the sized heaps to the
```
Comment (djwatson) [inline include/jemalloc/internal/arena_structs_b.h]: Not quite, you'd still want to allocate a new sized extent before using an unsized, yes?

File: include/jemalloc/internal/sized_region.h (hunk 1 above)
Comment (davidtgoldblatt) [reply include/jemalloc/internal/sized_region.h]: Fixed.

File: include/jemalloc/internal/arena_structs_b.h (hunk 2 above)
Comment (davidtgoldblatt) [reply include/jemalloc/internal/arena_structs_b.h]: Got rid of this whole comment and put a similar message in the extent code, where it makes more sense.

**Comment by davidtgoldblatt** [comment]
After last amend, run_tests.sh passes on both Linux and OS X. I tweaked down the 32-bit default size (we were running into VM exhaustion issues).

**Comment by djwatson** [comment]
I looked over the extent code fairly carefully since I just spent some time there, looks correct to me. 

Do we have prod numbers yet?

**Comment by davidtgoldblatt** [comment]
Working on it now.

**Comment by davidtgoldblatt** [comment]
Added some stats counters.

Initial prod benchmarks ranged from mild improvements to huge regressions. The regressions were so big I think something must be going wrong; digging into it now.

**Comment by interwq** [comment]
@davidtgoldblatt let's deprecate this one? Given the sdallocx integration (which should only get more popular over time), probably not worth paying the complexity here.

**Comment by davidtgoldblatt** [comment]
Sized regions, you were too beautiful for this world.

**Pull request closed without merging**